starlite/testing/async_test_client/transport.py

```diff
@@ -0,0 +1,187 @@
+        async def receive() -> "ReceiveMessage":
+            if context["request_complete"]:
+                if not context["response_complete"].is_set():
+                    await context["response_complete"].wait()
+                disconnect_event: "HTTPDisconnectEvent" = {"type": "http.disconnect"}
+                return disconnect_event
+
+            body = cast("Union[bytes, str, GeneratorType]", (request.read() or b""))
+            request_event: "HTTPRequestEvent" = {"type": "http.request", "body": b"", "more_body": False}
+            if isinstance(body, GeneratorType):  # pragma: no cover
+                try:
+                    chunk = body.send(None)
+                    request_event["body"] = chunk if isinstance(chunk, bytes) else chunk.encode("utf-8")
+                    request_event["more_body"] = True
+                except StopIteration:
+                    context["request_complete"] = True
+            else:
+                context["request_complete"] = True
+                request_event["body"] = body if isinstance(body, bytes) else body.encode("utf-8")
+            return request_event
+
+        return receive
+
+    @staticmethod
+    def create_send(request: "Request", context: SendReceiveContext) -> "Send":
+        async def send(message: "Message") -> None:
+            if message["type"] == "http.response.start":
+                if context["response_started"]:
+                    raise AssertionError('Received multiple "http.response.start" messages.')
+                context["raw_kwargs"]["status_code"] = message["status"]
+                context["raw_kwargs"]["headers"] = [
+                    (k.decode("utf-8"), v.decode("utf-8")) for k, v in message.get("headers", [])
+                ]
+                context["response_started"] = True
+            elif message["type"] == "http.response.body":
+                if not context["response_started"]:
+                    raise AssertionError('Received "http.response.body" without "http.response.start".')
+                if context["response_complete"].is_set():
+                    raise AssertionError('Received "http.response.body" after response completed.')
+                body = message.get("body", b"")
+                more_body = message.get("more_body", False)
+                if request.method != "HEAD":
+                    context["raw_kwargs"]["stream"].write(body)
+                if not more_body:
+                    context["raw_kwargs"]["stream"].seek(0)
+                    context["response_complete"].set()
+            elif message["type"] == "http.response.template":  # type: ignore[comparison-overlap] # pragma: no cover
+                context["template"] = message["template"]  # type: ignore[unreachable]
+                context["context"] = message["context"]
+
+        return send
+
+    def parse_request(self, request: "Request") -> Dict[str, Any]:
```

Comment:
again, there is a lot of duplication happening here. Cant this be refactored to use inheritence?

Reply:
this solves it [149a817](https://github.com/starlite-api/starlite/pull/980/commits/149a817b5cd84819582725c0ada2a9bbdd5d079d)